lib/validator/validator.cpp

```diff
@@ -3,32 +3,247 @@
+  }
+  void operator()(const StartSection &) {
+    // API:
+    // const Start &S = Sec.getContent();
+    // S.getFunctionIndex();
+    // S.getArguments();
+    // S.getResult();
+
+    // TODO: Validation requires f have functype with param arity and types
+    // matching arg* and result arity r.
+
+    // TODO: Validation appends the result types of f to the value index space
+    // (making them available for reference by subsequent definitions).
+
+    // TODO: In addition to the type-compatibility checks mentioned above, the
+    // validation rules for value definitions additionally require that each
+    // value is consumed exactly once. Thus, during validation, each value has
+    // an associated "consumed" boolean flag. When a value is first added to
+    // the value index space (via import, instance, alias or start), the flag
+    // is clear. When a value is used (via export, instantiate or start), the
+    // flag is set. After validating the last definition of a component,
+    // validation requires all values' flags are set.
+  }
+  void operator()(const ImportSection &Sec) {
+    // NOTE: This section share the validation rules with export section, one
+    // must share the implementation as much as possible.
+    for (const Import &I : Sec.getContent()) {
+      // TODO: Validation requires that annotated plainnames only occur on func
+      // imports or exports and that the first label of a [constructor],
+      // [method] or [static] matches the plainname of a preceding resource
+      // import or export, respectively, in the same scope (component, component
+      // type or instance type).
+
+      // TODO: Validation of [constructor] names requires that the func returns
+      // a (result (own $R)), where $R is the resource labeled r.
+
+      // TODO: Validation of [method] names requires the first parameter of the
+      // function to be (param "self" (borrow $R)), where $R is the resource
+      // labeled r.
+
+      // TODO: Validation of [method] and [static] names ensures that all field
+      // names are disjoint.
+      I.getName();
+
+      std::visit(ExternDescVisitor{}, I.getDesc());
+    }
+    // TODO: Validation requires that all resource types transitively used in
+    // the type of an export are introduced by a preceding importdecl or
+    // exportdecl.
+  }
+  void operator()(const ExportSection &Sec) {
+
+    for (const Export &E : Sec.getContent()) {
+      if (E.getDesc().has_value()) {
+        // TODO: Validation requires any exported sortidx to have a valid
+        // externdesc (which disallows core sorts other than core module). When
+        // the optional externdesc immediate is present, validation requires it
+        // to be a supertype of the inferred externdesc of the sortidx.
+        auto SI = E.getSortIndex();
+        SI.getSort();
+        SI.getSortIdx();
+        std::visit(ExternDescVisitor{}, *E.getDesc());
+      }
+    }
+  }
+
+private:
+  Validator *V;
```

Comment:
Use `const Validator &` here.

Reply:
I think `const Validator &` won't work, since
```c++
Expect<void> Validator::validate(const AST::Module &Mod)
```
is not a const method?